### PR TITLE
Add appendW, prependW and concatW to arrays

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Adds `appendW` and `prependW` to `Array` and `ReadonlyArray`, and `concatW` to `NonEmptyArray` and `ReadonlyNonEmptyArray`.

### CODE REVIEW

File: docs/modules/Array.ts.md [hunk 1]
```diff
@@ -1384,6 +1398,18 @@ assert.deepStrictEqual(pipe([2, 3, 4], prepend(1)), [1, 2, 3, 4])
...
 Added in v2.10.0
 
+## prependW
+
+Less strict version of [`prepend`](#prepend).
+
+**Signature**
+
+```ts
+export declare const prependW: <A, B>(head: B) => (tail: A[]) => NEA.NonEmptyArray<A | B>
```
Comment (thewilkybarkid) [inline docs/modules/Array.ts.md:1408]: Technically return type can be `[B, ...A[]]`, but suspect that's not particularly useful.

File: docs/modules/Array.ts.md [hunk 2]
```diff
@@ -1384,6 +1398,18 @@ assert.deepStrictEqual(pipe([2, 3, 4], prepend(1)), [1, 2, 3, 4])
...
+```
+
+Added in v2.11.0
```
Comment (thewilkybarkid) [inline docs/modules/Array.ts.md:1411]: I'm guessing.

File: src/NonEmptyArray.ts [hunk 3]
```diff
@@ -71,11 +71,21 @@ export const isOutOfBound = <A>(i: number, as: Array<A>): boolean => i < 0 || i
  */
 export const prepend = <A>(head: A) => (tail: Array<A>): NonEmptyArray<A> => [head, ...tail]
```
Comment (gcanti) [inline src/NonEmptyArray.ts:72]: This can be derived from `prependW` now

File: src/NonEmptyArray.ts [hunk 4]
```diff
@@ -71,11 +71,21 @@ export const isOutOfBound = <A>(i: number, as: Array<A>): boolean => i < 0 || i
...
 
+/**
+ * @internal
+ */
+export const prependW = <A, B>(head: B) => (tail: Array<A>): NonEmptyArray<A | B> => [head, ...tail]
+
 /**
  * @internal
  */
 export const append = <A>(end: A) => (init: Array<A>): NonEmptyArray<A> => concat(init, [end])
```
Comment (gcanti) [inline src/NonEmptyArray.ts:82]: This can be derived from `appendW` now

File: src/NonEmptyArray.ts [hunk 5]
```diff
@@ -225,6 +235,16 @@ export function concat<A>(first: Array<A>, second: Array<A>): Array<A> {
   return first.concat(second)
 }
 
+/**
+ * @category combinators
+ * @since 2.11.0
+ */
+export function concatW<A, B>(first: Array<A>, second: NonEmptyArray<B>): NonEmptyArray<A | B>
+export function concatW<A, B>(first: NonEmptyArray<A>, second: Array<B>): NonEmptyArray<A | B>
+export function concatW<A, B>(first: Array<A>, second: Array<B>): Array<A | B> {
```
Comment (gcanti) [inline src/NonEmptyArray.ts:244]: IMO this should go to `Array` and curried, see also (v3): https://github.com/gcanti/fp-ts/blob/ade368f43098a141fc67ccdb7b7fc9cf728e835c/src/ReadonlyArray.ts#L239-L257

**Review by gcanti — CHANGES_REQUESTED** [review]

File: src/ReadonlyNonEmptyArray.ts [hunk 6]
```diff
@@ -73,22 +73,23 @@ export const isOutOfBound = <A>(i: number, as: ReadonlyArray<A>): boolean => i <
 /**
  * @internal
  */
-export const prepend = <A>(head: A) => (tail: ReadonlyArray<A>): ReadonlyNonEmptyArray<A> => [head, ...tail]
+export const prepend = <A>(head: A) => (tail: ReadonlyArray<A>): ReadonlyNonEmptyArray<A> => prependW(head)(tail)
```
Comment (gcanti) [inline src/ReadonlyNonEmptyArray.ts:76]: Sorry, I wasn't clear, with "derived" I mean:

```ts
/**
 * @internal
 */
export const prependW = <B>(head: B) => <A>(tail: ReadonlyArray<A>): ReadonlyNonEmptyArray<A | B> => [head, ...tail]

/**
 * @internal
 */
export const prepend: <A>(head: A) => (tail: ReadonlyArray<A>) => ReadonlyNonEmptyArray<A> = prependW

/**
 * @internal
 */
export const appendW = <B>(end: B) => <A>(init: ReadonlyArray<A>): ReadonlyNonEmptyArray<A | B> => concatW([end])(init)

/**
 * @internal
 */
export const append: <A>(end: A) => (init: ReadonlyArray<A>) => ReadonlyNonEmptyArray<A> = appendW
```

File: src/NonEmptyArray.ts (hunk 5 above)
Comment (thewilkybarkid) [reply src/NonEmptyArray.ts:244]: I've had a go at backporting the changes from v3 and there seem to be type problems. Not sure what the difference is between TS 3.5 and 4.1 that seems to stop it from working. Going to back out of `concat` changes for this PR.

Comment (gcanti) [reply src/NonEmptyArray.ts]: @thewilkybarkid you are right, there's a type problem in v3 too, I'll look into that.

**Review by gcanti — APPROVED** [review]

File: src/NonEmptyArray.ts (hunk 5 above)
Comment (thewilkybarkid) [reply src/NonEmptyArray.ts]: Happy to have a go if you like.

Comment (gcanti) [reply src/NonEmptyArray.ts]: @thewilkybarkid fixed for now by exporting `concat` / `concatW` from [ReadonlyArray](https://github.com/gcanti/fp-ts/blob/86b8c14f78fb06a94f4337963332dfde74e563d4/src/ReadonlyArray.ts#L239-L250) and [ReadonlyNonEmptyArray](https://github.com/gcanti/fp-ts/blob/86b8c14f78fb06a94f4337963332dfde74e563d4/src/ReadonlyNonEmptyArray.ts#L204-L221)

**Pull request merged**